packages/grafana-ui/src/components/BarChart/BarChart.tsx

```diff
@@ -0,0 +1,310 @@
+        colorMode,
+        pathBuilder: config.drawBars,
+        pointsBuilder: config.drawPoints,
+        show: !customConfig.hideFrom?.graph,
+        gradientMode: customConfig.gradientMode,
+        thresholds: field.config.thresholds,
+
+        /*
+          lineColor: customConfig.lineColor ?? seriesColor,
+          lineWidth: customConfig.lineWidth,
+          lineStyle: customConfig.lineStyle,
+          */
+
+        // The following properties are not used in the uPlot config, but are utilized as transport for legend config
+        dataFrameFieldIndex: {
+          fieldIndex: i,
+          frameIndex: 0,
+        },
+        fieldName: getFieldDisplayName(field, data),
+        hideInLegend: customConfig.hideFrom?.legend,
+      });
+
+      // The builder will manage unique scaleKeys and combine where appropriate
+      builder.addScale({
+        scaleKey,
+        min: field.config.min,
+        max: field.config.max,
+        softMin: customConfig.axisSoftMin,
+        softMax: customConfig.axisSoftMax,
+        orientation: xOri === 0 ? 1 : 0,
+      });
+
+      if (customConfig.axisPlacement !== AxisPlacement.Hidden) {
+        let placement = customConfig.axisPlacement;
+        if (!placement || placement === AxisPlacement.Auto) {
+          placement = AxisPlacement.Left;
+        }
+        if (xOri === 1) {
+          if (placement === AxisPlacement.Left) {
+            placement = AxisPlacement.Bottom;
+          }
+          if (placement === AxisPlacement.Right) {
+            placement = AxisPlacement.Top;
+          }
+        }
+
+        builder.addAxis({
+          scaleKey,
+          label: customConfig.axisLabel,
+          size: customConfig.axisWidth,
+          placement,
+          formatValue: (v) => formattedValueToString(field.display!(v)),
+          theme,
+        });
+      }
+    }
+
+    return builder;
+  }, [data, configRev, orientation, width, height]);
+
+  const onLabelClick = useCallback(
+    (legend: VizLegendItem, event: React.MouseEvent) => {
+      const { fieldIndex } = legend;
+
+      if (!onLegendClick || !fieldIndex) {
+        return;
+      }
+
+      onLegendClick({
+        fieldIndex,
+        mode: GraphNGLegendEventMode.AppendToSelection,
+      });
+    },
+    [onLegendClick, data]
+  );
+
+  const hasLegend = useRef(legend && legend.displayMode !== LegendDisplayMode.Hidden);
+
+  const legendItems = configBuilder
+    .getSeries()
+    .map<VizLegendItem | undefined>((s) => {
+      const seriesConfig = s.props;
+      const fieldIndex = seriesConfig.dataFrameFieldIndex;
+      if (seriesConfig.hideInLegend || !fieldIndex) {
+        return undefined;
+      }
+
+      const field = data.fields[fieldIndex.fieldIndex];
+      if (!field) {
+        return undefined;
+      }
+
+      return {
+        disabled: !seriesConfig.show ?? false,
+        fieldIndex,
+        color: seriesConfig.lineColor!,
+        label: seriesConfig.fieldName,
+        yAxis: 1,
```

Comment:
Think we need to account for the axis being either left or right. See TimeSeries panel legend creation.

Then this function will become identical and I think we could turn it into some utility function `getGraphLegendItems(builder: UPlotConfigBuilder): VizLegendItem[]`

Reply:
I added this as a TODO on https://github.com/grafana/grafana/issues/30741